blaze-core/src/main/scala/org/http4s/blazecore/websocket/Serializer.scala

```diff
@@ -0,0 +1,117 @@
+
+/** Combined [[WriteSerializer]] and [[ReadSerializer]] */
+private trait Serializer[I] extends WriteSerializer[I] with ReadSerializer[I]
+
+/** Serializes write requests, storing intermediates in a queue */
+private trait WriteSerializer[I] extends TailStage[I] { self =>
+
+  ////////////////////////////////////////////////////////////////////////
+
+  private var serializerWriteQueue = new ArrayBuffer[I]
+  private var serializerWritePromise: Promise[Unit] = null
+
+  ///  channel writing bits //////////////////////////////////////////////
+  override def channelWrite(data: I): Future[Unit] =
+    channelWrite(data :: Nil)
+
+  override def channelWrite(data: Seq[I]): Future[Unit] = synchronized {
+    if (serializerWritePromise == null) {   // there is no queue!
+      serializerWritePromise = Promise[Unit]
+      val f = super.channelWrite(data)
+      f.onComplete(checkQueue)(directec)
+      f
+    } else {
+      serializerWriteQueue ++= data
+      serializerWritePromise.future
+    }
+  }
+
+  private def checkQueue(t: Try[Unit]): Unit = {
+    t match {
+      case f@ Failure(_) =>
+        val p = synchronized {
+          serializerWriteQueue.clear()
+          val p = serializerWritePromise
+          serializerWritePromise = null
+          p
+        }
+        p.tryComplete(f)
+        ()
+
+      case Success(_) =>
+        synchronized {
+          if (serializerWriteQueue.isEmpty) {
+            // Nobody has written anything
+            serializerWritePromise = null
+          } else {
+            // stuff to write
+            val f = {
+              if (serializerWriteQueue.length > 1) { // multiple messages, just give them the queue
+                val a = serializerWriteQueue
+                serializerWriteQueue = new ArrayBuffer[I](a.size + 10)
+                super.channelWrite(a)
+              } else {          // only a single element to write, don't send the while queue
+                val h = serializerWriteQueue.head
+                serializerWriteQueue.clear()
+                super.channelWrite(h)
+              }
+            }
+
+            val p = serializerWritePromise
+            serializerWritePromise = Promise[Unit]
+
+            f.onComplete { t =>
+              checkQueue(t)
+              p.complete(t)
+            }(trampoline)
+          }
+        }
+    }
+  }
+}
+
+/** Serializes read requests */
+trait ReadSerializer[I] extends TailStage[I] {
+  private val serializerReadRef = new AtomicReference[Future[I]](null)
+
+  ///  channel reading bits //////////////////////////////////////////////
+
+  override def channelRead(size: Int = -1, timeout: Duration = Duration.Inf): Future[I] = {
+    val p = Promise[I]
+    val pending = serializerReadRef.getAndSet(p.future)
+
+    if (pending == null) serializerDoRead(p, size, timeout)  // no queue, just do a read
+    else {
+      val started = if (timeout.isFinite()) System.currentTimeMillis() else 0
```

Comment:
Should this be using nanoTime to avoid the non-monotonic shenanigans of currentTimeMillis?  (I think yes).

Should this be using Deadline so you don't have to do this bookkeeping?  (I don't think you can, but maybe it clicks for you.)

Reply:
Deadline is much better and that is probably the better interface for blaze pipeline api's as well. Maybe they can make it into the blaze-0.14 release.

Reply:
The bad news about `Deadline` is that there aren't top or values for it. I'll go with making sure the times are monotonic.